stdlib/messaging/nats/src/main/java/org/ballerinalang/nats/basic/consumer/DefaultMessageHandler.java

```diff
@@ -102,36 +112,82 @@ private void dispatchWithDataBinding(ObjectValue msgObj, BType intendedType, byt
         try {
             Object typeBoundData = bindDataToIntendedType(data, intendedType);
             CountDownLatch countDownLatch = new CountDownLatch(1);
-            runtime.invokeMethodAsync(serviceObject, ON_MESSAGE_RESOURCE, new ResponseCallback(countDownLatch),
-                    msgObj, true, typeBoundData, true);
+            executeResource(msgObj, countDownLatch, typeBoundData);
             countDownLatch.await();
         } catch (NumberFormatException e) {
             ErrorValue dataBindError = Utils
                     .createNatsError("The received message is unsupported by the resource signature");
-            ErrorHandler.dispatchError(serviceObject, msgObj, dataBindError, runtime);
+            ErrorHandler.dispatchError(serviceObject, msgObj, dataBindError, runtime, connectedUrl);
+            NatsMetricsUtil.reportConsumerError(connectedUrl, msgObj.getStringValue(Constants.SUBJECT),
+                                                NatsObservabilityConstants.ERROR_TYPE_MSG_RECEIVED);
```

Comment:
Shall we remove metric publishing from here and publish error scenario in response callback notify success method